test_conformance/api/test_semaphores.cpp

```diff
@@ -0,0 +1,1137 @@
+    GET_PFN(deviceID, clEnqueueSignalSemaphoresKHR);
+    GET_PFN(deviceID, clEnqueueWaitSemaphoresKHR);
+    GET_PFN(deviceID, clReleaseSemaphoreKHR);
+
+    // Create ooo queue
+    clCommandQueueWrapper queue = clCreateCommandQueue(
+        context, deviceID, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &err);
+    test_error(err, "Could not create command queue");
+
+    // Create semaphores
+    cl_semaphore_properties_khr sema_props[] = {
+        static_cast<cl_semaphore_properties_khr>(CL_SEMAPHORE_TYPE_KHR),
+        static_cast<cl_semaphore_properties_khr>(CL_SEMAPHORE_TYPE_BINARY_KHR),
+        0
+    };
+    cl_semaphore_khr sema_1 =
+        clCreateSemaphoreWithPropertiesKHR(context, sema_props, &err);
+    test_error(err, "Could not create semaphore");
+
+    cl_semaphore_khr sema_2 =
+        clCreateSemaphoreWithPropertiesKHR(context, sema_props, &err);
+    test_error(err, "Could not create semaphore");
+
+    // Create user events
+    clEventWrapper user_event_1 = clCreateUserEvent(context, &err);
+    test_error(err, "Could not create user event");
+
+    clEventWrapper user_event_2 = clCreateUserEvent(context, &err);
+    test_error(err, "Could not create user event");
+
+    // Signal semaphore_1 (dependency on user_event_1)
+    clEventWrapper signal_1_event;
+    err = clEnqueueSignalSemaphoresKHR(queue, 1, &sema_1, nullptr, 1,
+                                       &user_event_1, &signal_1_event);
+    test_error(err, "Could not signal semaphore");
+
+    // Wait semaphore_1 and semaphore_2 (dependency on user_event_1)
+    clEventWrapper wait_event;
+    cl_semaphore_khr sema_list[] = { sema_1, sema_2 };
+    err = clEnqueueWaitSemaphoresKHR(queue, 2, sema_list, nullptr, 1,
+                                     &user_event_1, &wait_event);
+    test_error(err, "Could not wait semaphore");
+
+    // Signal semaphore_1 (dependency on wait_event and user_event_2)
+    clEventWrapper signal_2_event;
+    cl_event wait_list[] = { user_event_2, wait_event };
+    err = clEnqueueSignalSemaphoresKHR(queue, 1, &sema_1, nullptr, 2, wait_list,
+                                       &signal_2_event);
+    test_error(err, "Could not signal semaphore");
+
+    // Flush and delay
+    err = clFlush(queue);
+    test_error(err, "Could not flush queue");
+    std::this_thread::sleep_for(std::chrono::seconds(FLUSH_DELAY_S));
+
+    // Ensure all events are not completed
+    test_assert_event_inprogress(signal_1_event);
+    test_assert_event_inprogress(signal_2_event);
+    test_assert_event_inprogress(wait_event);
+
+    // Complete user_event_1 (expect failure as waiting on semaphore_2 is not
+    // allowed (unsignaled)
+    err = clSetUserEventStatus(user_event_1, CL_COMPLETE);
+    test_assert_error(err != CL_SUCCESS,
+                      "signal_2_event completed unexpectedly");
+
+    // Ensure signal_1 is completed while others failed (the second signal
+    // should fail as it depends on wait)
+    err = clFinish(queue);
```

Comment:
We have concerns about the use of clFinish() instead of calling clWaitForEvents(signal_1_event); On some implementations clFinish() could deadlock since some of the enqueued commands will not reach complete status until user_event_2 is signaled.

Reply:
ok, I will discuss this issue internally with my team